tests/lib/newlib/newlib_hook/src/main.c

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright (c) 2021 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <ztest.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <signal.h>
+#include <sys/stat.h>
+#include <sys/libc-hooks.h>
+
+char buf[8];
+
+/* define the function prototype for sbrk */
+void *sbrk(ptrdiff_t __incr);
+
+/**
+ * @brief Verify _open() stub in newlib libc-hooks.c
+ *
+ * @details The default weak function return -1.
+ */
+void test_newlib_stub_open(void)
+{
+	int fd;
+
+	fd = open("../prj.conf", O_RDWR | O_CREAT, 0664);
+	zassert_equal(fd, -1, "open an invalid fd");
+}
+
+/**
+ * @brief Verify _close() stub in newlib libc-hooks.c
+ *
+ * @details The default weak function return -1.
+ */
+void test_newlib_stub_close(void)
+{
+	int ret = 0;
+
+	ret = close(-1);
+	zassert_equal(ret, -1, "closed an invalid fd");
+}
+
+/**
+ * @brief Verify _getpid() stub in newlib libc-hooks.c
+ *
+ * @details The default weak function return 0.
+ */
+void test_newlib_stub_getpid(void)
+{
+	pid_t pid;
+
+	pid = getpid();
+	zassert_equal(pid, 0, "getpid failed");
+}
+
+/**
+ * @brief Verify _fstat() stub in newlib libc-hooks.c
+ *
+ * @details The default weak function return 0.
+ */
+void test_newlib_stub_fstat(void)
+{
+	struct stat status;
+	int ret;
+
+	ret = fstat(-1, &status);
+	zassert_equal(ret, 0, "fstat failed");
+}
+
+/**
+ * @brief Verify _kill() stub in newlib libc-hooks.c
+ *
+ * @details The default weak function return 0.
+ */
+void test_newlib_stub_kill(void)
+{
+	int ret;
+
+	ret = raise(SIGILL);
+	zassert_equal(ret, 0, "kill failed");
+}
+
+/**
+ * @brief Verify _read() stub in newlib libc-hooks.c
+ *
+ * @details Use read() functions to test _read().
+ */
+void test_newlib_stub_read(void)
+{
+#ifdef CONFIG_USERSPACE
+	/* Skip userspace because call z_zephyr_read_stdin() failure. */
+	ztest_test_skip();
+#else
+	int ret;
+
+	ret = read(-1, buf, 1);
+	zassert_not_equal(ret, 0, "read failed");
+#endif
+}
+
+/**
+ * @brief Verify _write() stub in newlib libc-hooks.c
+ *
+ * @details Use character output functions to test _write().
+ */
+void test_newlib_stub_write(void)
+{
+#ifdef CONFIG_USERSPACE
+	/* Skip userspace because call z_zephyr_write_stdout() failure. */
+	ztest_test_skip();
```

Comment:
Same here.

Reply:
There is still an another fatal error when call write() printf() putc() etc... .
Call read() is OK now. 
Could you tell me why this happen?

START - test_newlib_stub_write
E: Page fault at address 0x22 (error code 0x0)
E: Linear address not present in page tables
E: Access violation: supervisor thread not allowed to read
E: PTE: not present
E: EAX: 0x00000022, EBX: 0x00000022, ECX: 0x00000000, EDX: 0x00000001
E: ESI: 0x00000000, EDI: 0x00000000, EBP: 0x00138f3c, ESP: 0x00138f3c
E: EFLAGS: 0x00000003 CS: 0x0008 CR3: 0x00145000
E: call trace:
E: EIP: 0x0011180f
E:      0x00117813 (0x22)
E:      0x00116259 (0x6)
E:      0x0011702e (0x4)
E:      0x001173cb (0x4)
E:      0x00119bfb (0x3)
E:      0x0010ecda (0x3)
E:      corrupted? (bp=0x139e40)
E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
E: Current thread: 0x121080 (test_newlib_stub_write)
E: Halting system

Reply:
Did you properly configure all the user mode primitives (e.g. thread memory pool)?

Reply:
Thanks. After I call k_thread_system_pool_assign(k_current_get() in main and set CONFIG_HEAP_MEM_POOL_SIZE=1024 in prj.conf. The fatal error disappear on some platform, for example qemu_x86. But on some platform, there are also some error. So I removed the code(putc,printf,etc...), only keep write(). This does not affect the test.

Reply:
> The fatal error disappear on some platform, for example qemu_x86. But on some platform, there are also some error.

Could you elaborate? We should not be seeing any errors on any platforms.

Reply:
When call printf() putc() etc... in function test_newlib_stub_write in userspace. Run command "twister -T tests/lib/newlib/newlib_hook/" on my local environment, the result is "34 of 38 test configurations passed (89.47%), 4 failed, 32 skipped"
The 4 failed platform are qemu_cortex_r5 mps2_an385 qemu_arc_hs qemu_arc_em. The log listed below:
qemu_cortex_r5:
START - test_newlib_stub_write
aE: ***** DATA ABORT *****
E: Permission Fault @ 0x000230cc
E: r0/a1:  0x000230cc  r1/a2:  0x00020164  r2/a3:  0x000001b4
E: r3/a4:  0x000230cd r12/ip:  0x03fdcf34 r14/lr:  0x0000f919
E:  xpsr:  0x60000130
E: Faulting instruction address (r15/pc): 0x0000f280
E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
E: Current thread: 0x20180 (test_newlib_stub_write)
E: Halting system

mps2_an385:
START - test_newlib_stub_write
aE: ***** MPU FAULT *****
E:   Data Access Violation
E:   MMFAR Address: 0x2000518c
E: r0/a1:  0x2000518c  r1/a2:  0x20000164  r2/a3:  0x000001b4
E: r3/a4:  0x2000518d r12/ip:  0x0000c42f r14/lr:  0x0000b59d
E:  xpsr:  0x61000000
E: Faulting instruction address (r15/pc): 0x0000af08
E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
E: Current thread: 0x20000180 (test_newlib_stub_write)
E: Halting system

qemu_arc_hs:
START - test_newlib_stub_write
aE: ***** Exception vector: 0x6, cause code: 0x2, parameter 0x4
E: Address 0x80405000
E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
E: Current thread: 0x80401020 (test_newlib_stub_write)
E: Halting system
E: ***** Exception vector: 0x3, cause code: 0x0, parameter 0x0
E: Address 0x80001c3c
E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
E: Current thread: 0x80401020 (test_newlib_stub_write)
E: Halting system

qemu_arc_em:
START - test_newlib_stub_write
aE: ***** Exception vector: 0x6, cause code: 0x2, parameter 0x4
E: Address 0x80405000
E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
E: Current thread: 0x80401020 (test_newlib_stub_write)
E: Halting system
E: ***** Exception vector: 0x3, cause code: 0x0, parameter 0x0
E: Address 0x80001c60
E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
E: Current thread: 0x80401020 (test_newlib_stub_write)
E: Halting system

Reply:
In this testcase, I just test write stub.